lib/Arch/X86_C.cpp

```diff
@@ -0,0 +1,152 @@
+#include <vector>
+
+#include "anvill/Arch.h"
+#include "anvill/Decl.h"
+
+#include <glog/logging.h>
+#include <remill/Arch/Arch.h>
+
+namespace anvill {
+
+// Allocates the elements of the function signature of func to memory or
+// registers. This includes parameters/arguments, return values, and the return
+// stack pointer.
+void X86_C::AllocateSignature(FunctionDecl &fdecl, const llvm::Function &func) {
+  // Bind return values first to see if we have injected an sret into the
+  // parameter list. Then, bind the parameters. It is important that we bind the
+  // return values before the parameters in case we inject an sret.
+  bool injected_sret = false;
+  fdecl.returns = BindReturnValues(func, injected_sret);
+  fdecl.params = BindParameters(func, injected_sret);
+  BindReturnStackPointer(fdecl, func);
+}
+
+// The return stack pointer describes where the stack will be upon return from
+// the function in terms of the registers of the current function. For x86_C,
+// this is usually ESP + 4 since that is where the return address is stored.
+void X86_C::BindReturnStackPointer(
+      FunctionDecl &fdecl, const llvm::Function &func) {
+  // Check if the first argument is an sret. If it is, then by the X86_C ABI,
+  // the callee is responbile for returning said sret argument in %eax and
+  // cleaning up the sret argument with a `ret 4`. This changes the
+  // return_stack_pointer offset because it will now be 4 bytes higher than we
+  // thought.
+  //
+  // However, even if there is sret on the second argument as well, we do not
+  // need to worry about this. For some reason the callee is only responsible
+  // for cleaning up the case where an sret argument is passed in as the first
+  // argument.
+  if (func.hasParamAttribute(0, llvm::Attribute::StructRet)) {
+    fdecl.return_stack_pointer_offset = 8;
+  } else {
+    fdecl.return_stack_pointer_offset = 4;
+  }
+
```

Comment:
Below your might do `injected_sret`, though that doesn't seem to influence this code. Perhaps an additional `bool injected_sret` parameter should be passed in. Alternatively, you could have an integer value that you pass in, where the parameter and return value binding code is responsible for figuring out how much callee-cleanup needs to happen.

Reply:
Fix 1742dc9. Passed in injected_sret in this case.